ratis-docs/src/site/markdown/membership-change.md

```diff
@@ -0,0 +1,104 @@
+<!---
+  Licensed to the Apache Software Foundation (ASF) under one or more
+  contributor license agreements.  See the NOTICE file distributed with
+  this work for additional information regarding copyright ownership.
+  The ASF licenses this file to You under the Apache License, Version 2.0
+  (the "License"); you may not use this file except in compliance with
+  the License.  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-->
+
+# Ratis Membership Change
+
+## Overview
+
+Adjusting the configuration, such as replacing failed servers or changing replication levels, 
+is sometimes necessary in practice. 
+Ratis facilitates these functionalities through joint consensus, 
+allowing for simultaneous addition or replacement of multiple servers in a cluster.
+
+> For details on how joint consensus algorithm works, please refer to section 4.3 in 
+[Raft Paper](https://web.stanford.edu/~ouster/cgi-bin/papers/OngaroPhD.pdf)
+
+During membership changes, the availability guarantee of Raft is more vulnerable than usual.
+Agreement (for elections and entry commitment) requires separate majorities from both the
+old and new configurations.
+For example, when changing from a cluster of 3 servers to a different cluster of 9 servers, 
+agreement requires both 2 of the 3 servers in the old configuration 
+and 5 of the 9 servers in the new configuration.
+Be carefully to keep both separate majorities online! 
+
+## Adding a new server
+
+To add a new node (e.g., N3) to an existing group (e.g., N0, N1, N2), follow these steps:
+
+1. Start the new peer N3 with **EMPTY** group. 
+
+```java
+this.serverN3 = RaftServer.newBuilder()
+    .setGroup(RaftGroup.emptygroup())
+    .setProperties(properties)
+    .setServerId(N3)
+    .setStateMachine(userStateMachine)
+    .build();
+this.serverN3.start()
+```
```

Comment:
According to https://www.markdownguide.org/basic-syntax/#code-blocks-1 , it should indent with eight spaces
```java
-```java
-this.serverN3 = RaftServer.newBuilder()
-    .setGroup(RaftGroup.emptygroup())
-    .setProperties(properties)
-    .setServerId(N3)
-    .setStateMachine(userStateMachine)
-    .build();
-this.serverN3.start()
-```
+        RaftServer N3 = RaftServer.newBuilder()
+            .setGroup(RaftGroup.emptygroup())
+            .setProperties(properties)
+            .setServerId(n3id)
+            .setStateMachine(userStateMachine)
+            .build();
+        N3.start();
```

Reply:
Oh, I see.